### ISSUE TITLE
Low ROC AUC Score, High Validation Accuracy for Binary Classification

### ISSUE THREAD

**Issue by JJN123** [fchollet/keras#7110]
I compute an AUC ROC score every epoch as follows:
```
class Histories(keras.callbacks.Callback):
	
def __init__(self, validation_generator = None):
		super(Histories, self).__init__()
		self.validation_generator = validation_generator

	def on_train_begin(self, logs={}):
		self.aucs = []
		self.losses = []

	def on_epoch_end(self, epoch, logs={}):
		self.losses.append(logs.get('loss'))
		valid_steps = np.ceil(self.validation_generator.samples/self.validation_generator.batch_size)
		true_classes = self.validation_generator.classes
		predictions = self.model.predict_generator(generator = self.validation_generator, steps = valid_steps)

		self.aucs.append(roc_auc_score(y_true = true_classes, y_score = np.ravel(predictions)))
		print(self.aucs)
		return
```

The ROC AUC score's end up fluctuating around 0.5, for example:

[0.50508310249307475, 0.4822398100514444, 0.49138405223585269, 0.49789176889592401, 0.5143925603482391, 0.49754946576968734, 0.50938959240205783, 0.50715967550455088]             

For every model and data set combination I have tried running, including those which achieve high validation accuracy and low validation loss, and with balanced classes, I get these poor ROC AUC results. Without looking at more of my code, is there anything wrong with calculating an ROC AUC score as I have above? Training basically follows this tutorial: https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html

Note: roc_auc_score function is imported from sklearn

**Comment by jorgecarleitao** [comment]
AFAIK this is *not* related with Keras.

**Comment by JJN123** [comment]
I have now tested passing in test data stored as a numpy array called test_data. Using the following, I get an ROC AUC score of 0.85:

```
predictions = model.predict(validation_data, verbose =1)
roc_auc_score = roc_auc_score(validation_labels, predictions)
print('roc_auc_score from numpy array', roc_auc_score)
```


With the same model, same data, but locating the data in a directory, and passing this path into flow_from_directory, I get an ROC AUC score of 0.54:

```
test_datagen = ImageDataGenerator()

test_generator = test_datagen.flow_from_directory(
	        path_validation,
	        target_size=(img_width, img_height),
	        batch_size=batch_size,
	        class_mode='binary')

valid_steps = np.ceil(test_generator.samples/test_generator.batch_size)
true_classes = validation_generator.classes
predictions = model.predict_generator(generator = test_generator, steps = valid_steps)
print('auc score from generator', auc)
```

Any idea where this discrepancy comes from?

**Comment by haeseonii** [comment]
@JJN123 I am basically seeing the same behavior. I calculate all training, validation, and test AUC after very epoch (using ImageDataGenerator) and it basically oscillates between 0.5 while the training accuracy increases. Have you figured out what is going on?

**Comment by JJN123** [comment]
@haeseonii Yes I have figured out the issue. The default behavior of flow_from_directory is to shuffle the data; this means that the predictions from predict generator will not correspond to the ground truth labels which you get from test_generator.classes. So set shuffle = False.
